client/my-sites/plans-features-main/hooks/use-generate-action-hook.ts

```diff
@@ -0,0 +1,332 @@
+			) {
+				postButtonText = translate( "You've already used your free trial! Thanks!" );
+			}
+
+			return {
+				primary: {
+					callback: getActionCallback( { planSlug, cartItemForPlan, selectedStorageAddOn } ),
+					// TODO: Revisit status
+					status: 'enabled',
+					text,
+				},
+				postButtonText,
+			};
+		}
+
+		/**
+		 * 4. Logged-In (Admin) Plans actions
+		 */
+		let text = translate( 'Upgrade', { context: 'verb' } );
+		let status = null;
+		const current = sitePlanSlug === planSlug;
+		const isTrialPlan =
+			sitePlanSlug === PLAN_ECOMMERCE_TRIAL_MONTHLY ||
+			sitePlanSlug === PLAN_MIGRATION_TRIAL_MONTHLY ||
+			sitePlanSlug === PLAN_HOSTING_TRIAL_MONTHLY;
+
+		// TODO: Revisit WooExpress overrides and how to better structure this logic
+		let textOverride;
+
+		if ( isWooExpressMediumPlan( planSlug ) && ! isWooExpressMediumPlan( sitePlanSlug || '' ) ) {
+			textOverride = translate( 'Get Performance', { textOnly: true } );
+		} else if (
+			isWooExpressSmallPlan( planSlug ) &&
+			! isWooExpressSmallPlan( sitePlanSlug || '' )
+		) {
+			textOverride = translate( 'Get Essential', { textOnly: true } );
+		} else if ( isBusinessTrial( sitePlanSlug || '' ) ) {
+			textOverride = translate( 'Get %(plan)s', {
+				textOnly: true,
+				args: {
+					plan: getPlan( planSlug )?.getTitle() || '',
+				},
+			} );
+		}
+
+		if ( isFreePlan( planSlug ) ) {
+			text = translate( 'Contact support', { context: 'verb' } );
+
+			// TODO: Consider DRYing this up
+			if ( sitePlanSlug === planSlug && plansIntent !== 'plans-p2' ) {
+				text = translate( 'Manage add-ons', { context: 'verb' } );
+				status = 'enabled';
+			}
+		} else if (
+			availableForPurchase &&
+			sitePlanSlug &&
+			! current &&
+			! isTrialPlan &&
+			currentPlanBillingPeriod &&
+			billingPeriod &&
+			currentPlanBillingPeriod > billingPeriod
+		) {
+			text = translate( 'Contact support', { context: 'verb' } );
+		} else if (
+			availableForPurchase &&
+			sitePlanSlug &&
+			! current &&
+			getPlanClass( planSlug ) === getPlanClass( sitePlanSlug ) &&
+			! isTrialPlan
+		) {
+			// If the current plan matches on a lower-term, then show an "Upgrade to..." button.
+			if ( planMatches( planSlug, { term: TERM_TRIENNIALLY } ) ) {
+				text = textOverride || translate( 'Upgrade to Triennial' );
+			}
+
+			if ( planMatches( planSlug, { term: TERM_BIENNIALLY } ) ) {
+				text = textOverride || translate( 'Upgrade to Biennial' );
+			}
+
+			if ( planMatches( planSlug, { term: TERM_ANNUALLY } ) ) {
+				text = textOverride || translate( 'Upgrade to Yearly' );
+			}
+		} else if ( sitePlanSlug === planSlug && plansIntent !== 'plans-p2' ) {
+			// All other actions for a current plan
+			text = translate( 'View plan' );
+
+			if ( canUserManageCurrentPlan ) {
+				text = translate( 'Manage plan' );
+			} else if ( domainFromHomeUpsellFlow ) {
+				text = translate( 'Keep my plan', { context: 'verb' } );
+			}
+		} else if ( isStuck && availableForPurchase ) {
```

Comment:
If we can get rid of the `isStuck` context from here then I think we'd be much better off. It's a very grid-specific concept, and kinda contrary to what we're trying to do here (to generalise actions and detach the wording from the grid). 

Perhaps if we consider an alternative/second "text" to the primary action and use that? I guess challenge will be coming up with a proper name for a variable that will make sense... 🤔